src/osd/PG.cc

```diff
@@ -4234,7 +4234,7 @@ bool PG::sched_scrub()
   time_for_deep = (time_for_deep || deep_coin_flip);
 
   //NODEEP_SCRUB so ignore time initiated deep-scrub
-  if (osd->osd->get_osdmap()->test_flag(CEPH_OSDMAP_NODEEP_SCRUB) ||
+  if (get_osdmap()->test_flag(CEPH_OSDMAP_NODEEP_SCRUB) ||
```

Comment:
why do you think it's unsafe here? could you elaborate a little bit the racing we are facing without your change?

Reply:
we did see the issue that leading to segment fault when shard_ptr replacing, but it is on 12.2.7, which pubilsh_stats_to_osd() will call osd->osd->get_osdmap(). 

> (gdb) bt 
> #0 0x0000000000000000 in ?? ()
> #1 0x000055c38ab92549 in std::_Sp_counted_base<(_gnu_cxx::_Lock_policy)2>::_M_release() ()
> #2 0x000055c38ac76a76 in PG::publish_stats_to_osd() ()

std::shared_ptr is not thread safe

because std::shared_ptr is not thread safe, here get_osdmap() will read the OSDMapRef of class OSD without lock, so this behavior may race with write.

Reply:
i see. we update `OSD::osdmap` in `OSD::_committed_osd_maps()` which is executed by objectstore's finisher thread. while `PG::sched_scrub()` is called by OSD's sharded work queue's worker thread.

and we push the osdmap updates down to PGs `OSD::consume_map()` which is in turn called by `OSD::_committed_osd_maps()` where `osdmap` is updated. so it does not big deal if we are checking a stale  `CEPH_OSDMAP_NODEEP_SCRUB` flag. also this flag will be updated with the latest osdmap very soon.

@Aran85 since you are on this, could you update https://github.com/ceph/ceph/pull/26874/files#diff-dfb9ddca0a3ee32b266623e8fa489626L4247 as well?

Reply:
yes sir :D